### DISCUSSION TITLE
Reading numeric values of the SWIG object 'gp_XYZ'

### DISCUSSION THREAD

**Discussion by kfjordt** [IfcOpenShell/IfcOpenShell#2514] [Q&A]
Hello,

I am doing a project which requires parsing a lot of geometry. Therefore, I am using the geometry iterator. It contains the methods `compute_bounds()`, `bounds_min()`, `bounds_max()`, which all return a SWIG object of type gp_XYZ constant.

Is there any way to convert this data type to a float? 

The OpenCascade docs (https://dev.opencascade.org/doc/occt-6.9.1/refman/html/classgp___x_y_z.html) has methods for returning the x, y and z values of a gp_XYZ contant, but none of them works for me.

Lastly, is computing the bounds actually computationally faster than simply just creating the shape and calculating the min and max x,y,z values myself?

Thanks in advance!

**Comment by aothms** [comment]
`compute_bounds()` either evaluates the geometry completely in which case you'll evaluate it twice probably, once for calculating the bounds, the 2nd time applying that bounds center as an offset to center the model. Or it just looks at the placements in the model.

Secondly on SWIG, it's indeed not a fully automatic process. There are certain types you tell swig to "wrap" or one creates a "typemap" (custom conversion function) and the others remain opaque. This is the case here, there is no way to access the data from a gp_XYZ in python, not even with pythonOCC. (Although you could hack your way around it https://stackoverflow.com/questions/28904033/extracting-swig-wrapped-c-instance-pointer-for-use-in-cython)

So I think that would give you enough context, these are not widely used internal methods. In python it's equally performant and convenient to do it manually.

**Comment by Moult** [comment]
Referencing back to https://github.com/IfcOpenShell/IfcOpenShell/issues/1040 - would be really nice to get a set of functions for this. @aothms are you saying that derive_bbox() proposed in #1040 is equivalent to min/max(shape.verts) in terms of performance?

**Reply by aothms** [reply]
Yes, in theory it's possible to have more performant implementations for bbox calculation. But it would be like a completely parallel set of geometry implementation functions. Eg. for a polyhedral shell, you can do this by not looking at the faces at all, but just at the vertices (forgetting about the possibility that invalid faces would be removed from the output). For boolean union you can do a union of the bounding boxes. Boolean subtractions you can in some case ignore. Etc.

But for now (and likely never) something like this doesn't exist and therefore the sensible way to calculate a bbox is to do a min max of the evaluated triangulated mesh.
